### PR TITLE
Improve fuzzing coverage and improve processed data tracking

### LINKED ISSUE #2159: Unbuffered API: `DecryptError` is reported by TLS client when server uses large-sized certificates
**Describe the bug**
This issue can be found under the situation:
 - Server uses large certificate that require numbers of TLS messages to be sent to client.
 - One or more TLS messages are divided into multiple data packets.

After some investigations, the issue can be found after the patch set https://github.com/rustls/rustls/pull/2049 was merged. My guess is that this might be a regression issue related to the deframer or TLS message processing logic.

When a previous TLS message has been decrypted but not discarded, and the current TLS message is incomplete, upon receiving the full message, the `process_more_input()` will start back from the *decrypted content* that is not expected.

**To Reproduce**
Unfortunately, I can't reproduce the issue with existing examples because this issue is found under specific network channel (the TLS messages are divided).

**Applicable Version(s)**
v/0.23.13
v/0.23.14

**Additional context**
`UnbufferedServerConnection` and `UnbufferedClientConnection` are used in my case.

### PR DESCRIPTION
**Pull request by ctz**
This PR started as improving fuzzing, but adding working fuzzing of the unbuffered API involved redoing the fix to #2227, and that fixed #2159 at the same time.

The first commit can reach #2227 quite quickly, if cherry picked before that fix.

I will also reinstate the larger corpus repo once this is done.

fixes #2159

### CODE REVIEW

**Comment by djc** [comment]
(Already pushed up https://github.com/rustls/rustls-fuzzing-corpora earlier this week.)

**Comment by ctz** [comment]
> (Already pushed up https://github.com/rustls/rustls-fuzzing-corpora earlier this week.)

Ah, sweet, thanks.

File: fuzz/fuzzers/server.rs [hunk 1]
```diff
@@ -6,9 +6,47 @@ extern crate rustls;
 use std::io;
```
Comment (cpu) [inline fuzz/fuzzers/server.rs:6]: do the corpus files added in this commit have any interesting provenance?

File: fuzz/fuzzers/server.rs [hunk 2]
```diff
@@ -6,9 +6,47 @@ extern crate rustls;
...
 use std::sync::Arc;
 
-use rustls::server::ResolvesServerCert;
+use rustls::server::{Acceptor, ResolvesServerCert};
 use rustls::{ServerConfig, ServerConnection};
 
+fuzz_target!(|data: &[u8]| {
+    match data.split_first() {
```
Comment (cpu) [inline fuzz/fuzzers/server.rs:13]: would this be more effective dispatching on the MSB instead?

File: fuzz/fuzzers/unbuffered.rs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+use std::sync::Arc;
+
+use rustls::client::UnbufferedClientConnection;
+use rustls::server::{ResolvesServerCert, UnbufferedServerConnection};
+use rustls::unbuffered::{ConnectionState, UnbufferedStatus};
+use rustls::{ClientConfig, RootCertStore, ServerConfig, SideData};
+
+fuzz_target!(|data: &[u8]| {
+    let mut data = data.to_vec();
+    match data.split_first_mut() {
```
Comment (cpu) [inline fuzz/fuzzers/unbuffered.rs:15]: similar question here RE: MSB vs first byte.

File: fuzz/fuzzers/unbuffered.rs [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        Some((0x01, rest)) => server(rest),
+        Some((_, _)) | None => {}
+    }
+});
+
+fn client(data: &mut [u8]) {
+    let root_store = RootCertStore::empty();
+    let config = ClientConfig::builder()
+        .with_root_certificates(root_store)
+        .with_no_client_auth();
```
Comment (cpu) [inline fuzz/fuzzers/unbuffered.rs:26]: nit: inline `root_store`?

File: fuzz/fuzzers/unbuffered.rs [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    Client(UnbufferedClientConnection),
+    Server(UnbufferedServerConnection),
+}
+
+fn fuzz_unbuffered(mut data: &mut [u8], mut conn: ClientServer) {
+    while !data.is_empty() {
+        let consumed = match &mut conn {
+            ClientServer::Server(server) => {
+                let st = server.process_tls_records(data);
+                process(st)
```
Comment (cpu) [inline fuzz/fuzzers/unbuffered.rs:50]: nit: inline `st` here and on L53?

File: fuzz/fuzzers/unbuffered.rs [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        };
+    }
+}
+
+fn process<S: SideData>(status: UnbufferedStatus<'_, '_, S>) -> Option<usize> {
+    let UnbufferedStatus { discard, state } = status;
+
+    match state {
+        Ok(ConnectionState::EncodeTlsData(mut enc)) => {
+            let mut buffer = [0u8; 16384];
```
Comment (cpu) [inline fuzz/fuzzers/unbuffered.rs:72]: is this buffer size significant in some way? Maybe the literal could be written `16_384`?

File: fuzz/fuzzers/unbuffered.rs [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+            enc.encode(&mut buffer).unwrap();
+        }
+        Ok(ConnectionState::TransmitTlsData(xmit)) => xmit.done(),
+        Ok(ConnectionState::WriteTraffic(_)) => return None,
+        Ok(ConnectionState::BlockedHandshake) => return None,
+        Ok(st) => panic!("unhandled state {st:?}"),
+        Err(_err) => return None,
```
Comment (cpu) [inline fuzz/fuzzers/unbuffered.rs:79]: nit: `Err(_)`?

File: fuzz/fuzzers/unbuffered.rs [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    };
+
+    Some(discard)
+}
+
+#[derive(Debug)]
+struct Fail;
```
Comment (cpu) [inline fuzz/fuzzers/unbuffered.rs:86]: nit: more descriptive name? `NoneCertResolver`?

**Review by cpu — APPROVED** [review]
Nice, :+1:

File: fuzz/fuzzers/server.rs (hunk 1 above)
Comment (ctz) [reply fuzz/fuzzers/server.rs:6]: They are copies of the existing ones, with 0x00 or 0x01 prepended, so they reach `fuzz_buffered_api` and `fuzz_acceptor_api`.

File: fuzz/fuzzers/server.rs (hunk 2 above)
Comment (cpu) [reply fuzz/fuzzers/server.rs:13]: This comment falls away with my understanding from [the other thread](https://github.com/rustls/rustls/pull/2249#discussion_r1869969109).

File: fuzz/fuzzers/unbuffered.rs (hunk 3 above)
Comment (cpu) [reply fuzz/fuzzers/unbuffered.rs:15]: This comment falls away with my understanding from [the other thread](https://github.com/rustls/rustls/pull/2249#discussion_r1869969109).

File: fuzz/fuzzers/unbuffered.rs (hunk 6 above)
Comment (ctz) [reply fuzz/fuzzers/unbuffered.rs]: fixed; should actually be `16_384 + 5`. Maybe in the future we could export this as a const.

**Pull request merged**